vault_renewal.go

```diff
@@ -2,129 +2,141 @@ package main
+			currentDatabaseCredentialsLease = databaseCredentialsLease
 		}
 	}
 }
 
-// renewUntilMaxTTL is a blocking helper function that uses LifetimeWatcher to
-// periodically renew the given secret or token when it is close to its
-// 'token_ttl' lease expiration time until it reaches its 'token_max_ttl' lease
-// expiration time.
-func (v *Vault) renewUntilMaxTTL(ctx context.Context, secret *vault.Secret, label string) error {
-	/* */ log.Printf("%s renew cycle: started; lease duration: %ds", label, secret.LeaseDuration)
-	defer log.Printf("%s renew cycle: the secret can no longer be renewed", label)
+type renewResult uint8
+
+const (
+	renewError renewResult = 1 << iota
+	exitRequested
+	expiringAuthToken                // will be revoked soon
+	expiringDatabaseCredentialsLease // will be revoked soon
+)
+
+// renewLeases is a blocking helper function that uses LifetimeWatcher
+// instances to periodically renew the given secrets when they are close to
+// their 'token_ttl' expiration times until one of the secrets is close to its
+// 'token_max_ttl' lease expiration time.
+func (v *Vault) renewLeases(ctx context.Context, authToken, databaseCredentialsLease *vault.Secret) (renewResult, error) {
+	/* */ log.Println("renew cycle: begin")
+	defer log.Println("renew cycle: end")
+
+	// auth token
+	authTokenWatcher, err := v.client.NewLifetimeWatcher(&vault.LifetimeWatcherInput{
+		Secret: authToken,
+	})
+	if err != nil {
+		return renewError, fmt.Errorf("unable to initialize auth token lifetime watcher: %w", err)
+	}
+
+	go authTokenWatcher.Start()
+	defer authTokenWatcher.Stop()
 
-	watcher, err := v.client.NewLifetimeWatcher(&vault.LifetimeWatcherInput{
-		Secret: secret,
+	// database credentials
+	databaseCredentialsWatcher, err := v.client.NewLifetimeWatcher(&vault.LifetimeWatcherInput{
+		Secret: databaseCredentialsLease,
 	})
 	if err != nil {
-		return fmt.Errorf("unable to initialize %s lifetime watcher: %w", label, err)
+		return renewError, fmt.Errorf("unable to initialize database credentials lifetime watcher: %w", err)
 	}
 
-	go watcher.Start()
-	defer watcher.Stop()
+	go databaseCredentialsWatcher.Start()
+	defer databaseCredentialsWatcher.Stop()
 
+	// monitor events from both watchers
 	for {
 		select {
 		case <-ctx.Done():
-			return context.Canceled
+			return exitRequested, nil
 
 		// DoneCh will return if renewal fails, or if the remaining lease
 		// duration is under a built-in threshold and either renewing is not
 		// extending it or renewing is disabled.  In both cases, the caller
 		// should attempt a re-read of the secret. Clients should check the
 		// return value of the channel to see if renewal was successful.
-		case err := <-watcher.DoneCh():
-			if err != nil {
-				return fmt.Errorf("%s renewal failed: %w", label, err)
-			}
+		case err := <-authTokenWatcher.DoneCh():
+			// Leases created by a token get revoked when the token is revoked.
+			return expiringAuthToken | expiringDatabaseCredentialsLease, err
```

Comment:
I'm also not familiar with the use of this single pipe here, please enlighten me!

Reply:
will explain below